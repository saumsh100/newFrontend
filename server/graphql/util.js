
import { getOffsetWithDefault, offsetToCursor } from 'graphql-relay';

/**
 * A simple function that accepts an array and connection arguments, and returns
 * a connection object for use in GraphQL
 *
 * This is a fork of connectionFromArraySlice
 * from https://github.com/graphql/graphql-relay-js/blob/master/src/connection/arrayconnection.js#L65
 *
 * Diferente form the original, is this case we build the cursor and pageInfo based on args
 * whitout slicing the array.
 * That way we can pass the already sliced or filtered array to the function.
 *
 * For now, this only works with forward pagination.
 */
// eslint-disable-next-line import/prefer-default-export
export function connectionFromArrayWithoutSlice(array, args, meta) {
  const { after, first, last } = args;
  const { arrayLength } = meta;

  if (typeof first === 'number') {
    if (first < 0) {
      throw new Error('Argument "first" must be a non-negative integer');
    }
  }
  if (typeof last === 'number') {
    if (last < 0) {
      throw new Error('Argument "last" must be a non-negative integer');
    }
  }

  const afterOffset = getOffsetWithDefault(after, 0);

  const startOffset = Math.max(afterOffset, -1) + 1;
  const endOffset = startOffset + first;

  const edges = array.map((value, index) => ({
    cursor: offsetToCursor(startOffset + index),
    node: value,
  }));

  const firstEdge = edges[0];
  const lastEdge = edges[edges.length - 1];
  const lowerBound = after ? afterOffset + 1 : 0;
  const upperBound = arrayLength;

  return {
    edges,
    pageInfo: {
      startCursor: firstEdge ? firstEdge.cursor : null,
      endCursor: lastEdge ? lastEdge.cursor : null,
      hasPreviousPage: typeof last === 'number' ? startOffset > lowerBound : false,
      hasNextPage: typeof first === 'number' ? endOffset < upperBound : false,
    },
  };
}
