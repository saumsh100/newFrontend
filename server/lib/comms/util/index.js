
import groupBy from 'lodash/groupBy';
import map from 'lodash/map';
import toArray from 'lodash/toArray';

const ATTRS = {
  email: 'email',
  sms: 'mobilePhoneNumber',
  phone: 'mobilePhoneNumber',
};

const ERROR_CODES = {
  email: '1100',
  sms: '1200',
  phone: '1300',
};

const PREFERENCE_TYPES = {
  email: 'emailNotifications',
  sms: 'sms',
  phone: 'phone',
};

const NO_PREFERENCE_ERROR_CODES = {
  email: '2100',
  sms: '2200',
  phone: '2300',
};

/**
 * cannotSend
 *
 * @param patient
 * @param primaryType
 * @returns {{errors: Array, success: Array}}
 */
export function cannotSend(patient, primaryType) {
  // If it is undefined return the error code
  if (!patient[ATTRS[primaryType]]) {
    return ERROR_CODES[primaryType];
  }
}

/**
 * noPreference
 *
 * @param patient
 * @param primaryType
 * @returns {{errors: Array, success: Array}}
 */
export function noPreference(patient, primaryType) {
  const preferences = patient.preferences || {};

  // If it is false return errorCode
  if (!preferences[PREFERENCE_TYPES[primaryType]]) {
    return NO_PREFERENCE_ERROR_CODES[primaryType];
  }
}

/**
 * generateOrganizedPatients will accept an array of patients and an
 * array of primaryTypes and will return the errors and successes based
 * on patientData. The point is to organize the failed responses so that
 * certain jobs can save a bulk amount to save time
 *
 * @param patients
 * @param primaryTypes (ie.// ['email', 'sms'])
 * @returns {{errors: Array, success: Array}}
 */
export function generateOrganizedPatients(patients, primaryTypes) {
  const errors = [];
  const success = [];

  for (const patient of patients) {
    primaryTypes.forEach((primaryType) => {
      const errorCode = cannotSend(patient, primaryType) || noPreference(patient, primaryType);
      if (errorCode) {
        errors.push({
          patient,
          errorCode,
          primaryType,
        });
      } else {
        success.push({ primaryType, patient });
      }
    });
  }

  return { errors, success };
}

/**
 * organizeForOutbox is a function that accepts an array and a selectorPredicate
 * and will group the array based on the selectorPredicate, and then map over that array
 * to return outbox items that have primaryTypes grouped together.
 *
 * NOTE: this is primarily a layer ontop of the mapPatientsTo____ calls that will take the flat
 * array with primaryType and bring it back into primaryTypes
 *
 * @param array (ie.// [{ patient, appointment, primaryType: 'sms' }, { patient, appointment, primaryType: 'email' }])
 * @param selectorPredicate (ie.// ({ patient, appointment }) => `${patient.id}_${appointment.id}`)
 * @param mergePredicate
 * @returns [organizedArray] (ie.// [{ patient, appointment, primaryTypes: ['sms', 'email'] }])
 */
export function organizeForOutbox(array, selectorPredicate, mergePredicate) {
  const groupedMap = groupBy(array, selectorPredicate);
  return toArray(map(groupedMap, mergePredicate));
}
