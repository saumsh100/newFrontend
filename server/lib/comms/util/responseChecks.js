
import find from 'lodash/find';

const THE_MAGIC_WORDS = {
  C: true,
  Y: true,
};

/**
 * getIndicesOf will return the indices in the text
 * of all the magicWord
 *
 * @param text
 * @param magicWord
 * @returns {Array}
 */
export function getIndicesOf(text, magicWord) {
  const magicWordLen = magicWord.length;
  if (magicWordLen === 0) {
    return [];
  }

  const indices = [];
  const regexp = new RegExp(magicWord, 'g');

  let match;
  while (match = regexp.exec(text))  {
    indices.push(match.index);
  }

  return indices;
}

/**
 *
 * @param char
 */
export function isWordy(char) {
  return /[a-zA-Z0-9'"]/.test(char);
}

/**
 *
 * @param text
 * @param magicWord
 */
export function isNotInWord(text, magicWord) {
  // If exactly equal to magicWord then don't go any further
  if (text === magicWord) {
    return true;
  }

  // Get all indicies of the magicWord in the text
  const indices = getIndicesOf(text, magicWord);

  // Loop through every occurance and see if any of the before character or
  // after character is alpha-numeric. If so, then return false
  for (const index of indices) {
    const beforeCharacter = text[index - 1];
    const afterCharacter = text[index + 1];
    if (!beforeCharacter && !isWordy(afterCharacter)) {
      return true;
    }

    if (!afterCharacter && !isWordy(beforeCharacter)) {
      return true;
    }

    if (beforeCharacter && !isWordy(beforeCharacter) && afterCharacter && !isWordy(afterCharacter)) {
      return true;
    }
  }

  return false;
}

/**
 * Determines if a certain string of text is a valid confirmation response
 *
 * @param text
 * @param magicWord
 * @returns {boolean}
 */
export function wordCheck(text, magicWord) {
  // If exactly equal to magicWord then don't go any further
  if (text === magicWord) {
    return true;
  }

  // If the text doesn't have the magic word at all, don't bother continuing!
  const hasMagicWord = text.includes(magicWord);

  // If the magicWord is part of another word, it doesn't count
  // ex.// magicWord='C'  text='Can't make it'
  const magicWordIsNotInWord = isNotInWord(text, magicWord);

  // This function plays nicer with checking AND for all booleans
  return hasMagicWord && magicWordIsNotInWord;
}

/**
 * isSmsConfirmationResponse determines if a certain string of text is
 * a valid confirmation response
 *
 * @param text
 * @param magicWords
 * @returns {boolean}
 */
export function isSmsConfirmationResponse(text, magicWords = THE_MAGIC_WORDS) {
  text = text.toUpperCase();

  return typeof magicWords !== 'string'
    ? !!find(magicWords, (value, key) => wordCheck(text, key))
    : wordCheck(text, magicWords);
}
