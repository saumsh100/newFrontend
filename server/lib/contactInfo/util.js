
const orderBy = require('lodash/orderBy');

/**
 * isHead checks if a given patient is the head of the given family by comparing accountId,
 * familyId and headId.
 * @param family
 * @param patient
 * @return {boolean}
 */
const isHead = (family, patient) =>
  !!family &&
  !!patient &&
  family.accountId === patient.accountId &&
  family.id === patient.familyId &&
  family.headId === patient.id;

/**
 * isActive checks if a patient has status equals to the string 'Active'
 * @param patient
 * @return {boolean}
 */
const isActive = ({ status }) => status === 'Active';

/**
 * getFromNewestCreatedFamily orders the a list of patients by its family creation date and
 * grabs the first.
 * @param patients
 */
const getFromNewestCreatedFamily = patients =>
  orderBy(
    patients,
    ({ family: { pmsCreatedAt, createdAt } }) => pmsCreatedAt || createdAt,
    'desc',
  )[0];

/**
 * getOldest orders a list of patients by their birth dates and grabs the first, thus the oldest.
 * @param patients
 */
const getOldest = patients => orderBy(patients, p => p.birthDate, 'asc')[0];

/**
 * getPatientAge is a utility function to extract the patient age form its birth date by year
 * diffing. This was created using plain JS so we can reduce the usage of moment.js.
 * @param patient
 * @return {number}
 */
const getPatientAge = (patient) => {
  const ageDate = new Date(Date.now() - new Date(patient.birthDate).getTime());
  return Math.abs(ageDate.getUTCFullYear() - 1970);
};

/**
 * adultChecker is a configurable function builder that creates a function that will check if an age
 * of a patient is within the adult range provided to it.
 * @param min
 * @param max
 * @return {function(*=): boolean}
 */
const adultChecker = (min, max) => patient =>
  min < getPatientAge(patient) && getPatientAge(patient) < max;

/**
 * grabPoC is a function that takes a patient list and returns a function that filters this list by
 * the given predicate, in case of the filter returns more than one result the fallbackHandler is
 * applied to the list.
 * This function use partially applied arguments so we can reuse the same list of patients but with
 * different predicates and fallbackHandlers.
 * @param patients
 * @return {Function}
 */
const grabPoC = patients => (filterPredicate, fallbackHandler) => {
  const filteredPatients = patients.filter(filterPredicate);
  if (filteredPatients.length === 1) {
    return filteredPatients[0];
  } else if (filteredPatients.length > 1) {
    return fallbackHandler(filteredPatients);
  }

  return undefined;
};

module.exports = {
  adultChecker,
  grabPoC,
  isHead,
  isActive,
  getFromNewestCreatedFamily,
  getOldest,
};
