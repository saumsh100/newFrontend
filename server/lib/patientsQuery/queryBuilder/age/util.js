
import { setDateToTimezone } from '@carecru/isomorphic';

/**
 * Compose a query-builder-ready parameters
 * @param params
 * @return {*}
 */
export function prepareQueryParams(params) {
  if (Array.isArray(params)) {
    return calculateDatesArray(params);
  }

  const reformattedParameters = Object.entries(params).map(([key, value]) => {
    const formattedValue = Array.isArray(value)
      ? calculateDatesArray(value)
      : calculateDate(value);
    return { [key]: formattedValue };
  });

  return Object.assign(...reformattedParameters);
}

/**
 * Calculate the date difference for an array of years.
 * @param {Array} yearValue
 * @return {string[]}
 */
function calculateDatesArray(yearValue) {
  return yearValue.map(calculateDate).sort();
}

/**
 * Calculate a difference in years, by subtracting the year given as parameters from the
 * current date.
 * @param years
 * @return {string}
 */
export function calculateDate(years) {
  const now = setDateToTimezone()
    .hours(0)
    .minutes(0)
    .seconds(0)
    .milliseconds(0);

  return now.subtract(years, 'years').toISOString();
}
