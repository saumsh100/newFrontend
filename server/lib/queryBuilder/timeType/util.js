
import { time } from '@carecru/isomorphic';

/**
 * Normalize a whole request object to have a correct date.
 * @param requestObject
 * @return {*}
 */
const normalizeRequest = (requestObject) => {
  if (typeof requestObject === 'string') {
    return normalizeTime(requestObject);
  }
  if (Array.isArray(requestObject)) {
    return requestObject.map(normalizeTime);
  }

  const normalizedObjects = Object.entries(requestObject)
    .map(([key, value]) => ({ [key]: normalizeRequest(value) }));

  return Object.assign(...normalizedObjects);
};

/**
 * Create a complete date-time from the requested time value.
 * @param requestTime
 * @return {string}
 */
const normalizeTime = (requestTime) => {
  const [hours, minutes] = requestTime.split(':');
  return time(hours, minutes).toISOString();
};


export default normalizeRequest;
