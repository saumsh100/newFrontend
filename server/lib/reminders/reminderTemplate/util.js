
import { sortAsc, setDateToTimezone } from '@carecru/isomorphic';
import { getReminderType } from '../createReminderText';
import isFeatureFlagEnabled from '../../featureFlag';

export function generateFamilyDetails(dependants, timezone) {
  if (dependants.length > 1) {
    const familyDetails = dependants
      .sort(
        (
          { appointment: { startDate: a } },
          { appointment: { startDate: b } },
        ) => sortAsc(a, b),
      )
      .map(({ appointment: { startDate }, firstName, lastName }) => {
        const { date, time } = getDateAndTime(startDate, timezone);
        return `\n${firstName} ${lastName}\n${date} at ${time}\n`;
      })
      .join('');

    return { familyDetails };
  }

  const [familyMember] = dependants;
  const {
    appointment: { startDate },
  } = familyMember;

  return {
    familyMember,
    ...getDateAndTime(startDate, timezone),
  };
}

export function getDateAndTime(date, timezone) {
  const mDate = setDateToTimezone(date, timezone);
  return {
    date: mDate.format('MMMM Do'),
    time: mDate.format('h:mma'),
  };
}

/**
 * Feature flag is here so that both the Preview API and the actual
 * reminders job can utilize the same function
 * @param account
 * @return {Promise<boolean>}
 */
export async function shouldUseLegacyTemplate(account) {
  const isCustomizable = isFeatureFlagEnabled(
    'customizable-reminders-templates',
    false,
    {
      userId: 'carecru-api',
      accountId: account.id,
      enterpriseId: account.enterpriseId,
    },
  );

  return !isCustomizable;
}

export function getReminderTemplateType({
  account,
  appointment,
  reminder,
  currentDate,
  dependants,
}) {
  if (isFamily(dependants)) {
    if (appointment) {
      return 'self';
    }

    return dependants.length === 1 ? 'single' : 'multiple';
  }

  const templateTypes = {
    weeksAway: 'weeks-away',
    weekAway: 'week-away',
    sameWeek: 'same-week',
    sameDay: 'same-day',
  };

  // Don't pass in appointment so it uses the reminder.interval instead
  const reminderType = getReminderType({
    account,
    appointment,
    reminder,
    currentDate,
  });

  return templateTypes[reminderType];
}

export function isFamily(dependants) {
  return dependants && dependants.length > 0;
}
