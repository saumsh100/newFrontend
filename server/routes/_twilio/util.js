
import mapKeys from 'lodash/mapKeys';

export function snakeCaseToCamelCase(str) {
  const find = /(\_\w)/g;
  const convert =  matches => matches[1].toUpperCase();
  return str.replace(
    find,
    convert
  );
}

export function convertKeysToCamelCase(obj) {
  return mapKeys(obj, (val, key) => {
    const firstLetter = key.charAt(0).toLowerCase();
    const removedPascal = firstLetter + key.slice(1, key.length);
    return snakeCaseToCamelCase(removedPascal);
  });
}

export function sanitizeTwilioSmsData(twilioData) {
  const cleansedData = convertKeysToCamelCase(twilioData);

  // Easily parse mediaData
  let mediaData = {};
  const numMedia = parseInt(cleansedData.numMedia);
  if (numMedia) {
    for (let i = 0; i < numMedia; i++) {
      mediaData[i] = {
        url: cleansedData[`mediaUrl${i}`],
        contentType: cleansedData[`mediaContentType${i}`],
      };
    }
  }

  return Object.assign({}, cleansedData, {
    id: cleansedData.sid || cleansedData.messageSid,
    mediaData,
  });
}
