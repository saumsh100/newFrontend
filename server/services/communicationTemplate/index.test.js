
import { AccountTemplate, Template } from 'CareCruModels';
import { seedTestUsers, accountId } from '../../../tests/util/seedTestUsers';
import { wipeAllModels } from '../../../tests/util/wipeModel';
import { omitProperties } from '../../../tests/util/selectors';
import { findAccountTemplate, validateParameters, composeMessage, getMessageFromTemplates } from './';

const templateId1 = '0486edfc-7e4f-464f-8fe7-a38bb6a37a52';
const templateId2 = '1486edfc-7e4f-464f-8fe7-a38bb6a37a52';
const templateId3 = '1486edfc-7e4f-464f-8fe7-a38bb6a37a53';
const accountTemplateId1 = '2486edfc-7e4f-464f-8fe7-a38bb6a37a52';
const accountTemplateId2 = '3486edfc-7e4f-464f-8fe7-a38bb6a37a52';
const accountTemplateId3 = '4486edfc-7e4f-464f-8fe7-a38bb6a37a52';
const accountTemplateId4 = '5486edfc-7e4f-464f-8fe7-a38bb6a37a52';

describe('services.communicationTemplate', () => {
  beforeAll(async () => {
    await seedTestUsers();
    await Template.bulkCreate([
      {
        id: templateId1,
        templateName: 'test-template',
        values: { name: true },
      },
      {
        id: templateId2,
        templateName: 'test-template-1',
        values: {
          age: true,
          name: true,
        },
      },
      {
        id: templateId3,
        templateName: 'test-template-2',
        values: {
          clinic: true,
          'patient.firstName': true,
          'patient.lastName': true,
          'patient.age': true,
        },
      },
    ]);
    await AccountTemplate.bulkCreate([
      {
        id: accountTemplateId1,
        templateId: templateId1,
        value: 'Great One ${name}',
      },
      {
        id: accountTemplateId2,
        templateId: templateId2,
        value: 'default One ${name}, ${age}.',
      },
      {
        id: accountTemplateId4,
        templateId: templateId3,
        value: 'Welcome ${patient.firstName} ${patient.lastName}, from ${clinic}.',
      },
      {
        id: accountTemplateId3,
        accountId,
        templateId: templateId2,
        value: 'Personalized template for ${name}, ${age}.',
      },
    ]);
  });

  afterAll(async () => {
    await wipeAllModels();
  });

  describe('DB related tests', () => {
    it('finds the accountTemplate for a given accountId and templateName', async () => {
      const accountTemplate = await findAccountTemplate(accountId, 'test-template-1');
      const rawTemplate = accountTemplate.get({ plain: true });
      const omitedTemplate = omitProperties(rawTemplate, ['template']);
      const omitedRawTemplate = omitProperties(rawTemplate.template);
      expect(omitedTemplate).toMatchSnapshot();
      expect(omitedRawTemplate).toMatchSnapshot();
    });

    it('returns the default accountTemplate if account specific template is not found', async () => {
      const accountTemplate = await findAccountTemplate(accountId, 'test-template');
      const rawTemplate = accountTemplate.get({ plain: true });
      const omitedTemplate = omitProperties(rawTemplate, ['template']);
      const omitedRawTemplate = omitProperties(rawTemplate.template);
      expect(omitedTemplate).toMatchSnapshot();
      expect(omitedRawTemplate).toMatchSnapshot();
    });
  });

  describe('template compiler', () => {
    it('validates template params', () => {
      const templateParams = {
        name: true,
        age: true,
      };
      const givenParams = {
        name: 'Donna',
        age: 22,
        address: 'My city',
      };

      expect(validateParameters(templateParams, givenParams)).toMatchSnapshot();
    });

    it('validates template params for object childs', () => {
      const templateParams = {
        name: true,
        age: true,
        'patient.firstName': true,
      };
      const givenParams = {
        name: 'Donna',
        age: 22,
        address: 'My city',
        patient: { firstName: 'Mike' },
      };

      expect(validateParameters(templateParams, givenParams)).toMatchSnapshot();
    });

    it('assign parameters', () => {
      const data = {
        name: 'Donna',
        age: 22,
      };

      expect(composeMessage('cool string, ${age}, ${name}.', data)).toBe('cool string, 22, Donna.');
    });
  });

  it('compiles the message with a default template', async () => {
    const message = await getMessageFromTemplates(accountId, 'test-template', {
      name: 'Donna',
      age: 22,
    });
    expect(message).toBe('Great One Donna');
  });

  it('compiles the message with the account-specified template', async () => {
    const message = await getMessageFromTemplates(accountId, 'test-template-1', {
      name: 'Donna',
      age: 22,
    });
    expect(message).toBe('Personalized template for Donna, 22.');
  });

  it('compiles the message with nasted object params', async () => {
    const data = {
      clinic: 'Donna dental',
      patient: {
        firstName: 'Michael',
        lastName: 'Myers',
        age: '40',
      },
    };
    const message = await getMessageFromTemplates(accountId, 'test-template-2', data);
    expect(message).toBe('Welcome Michael Myers, from Donna dental.');
  });
});
