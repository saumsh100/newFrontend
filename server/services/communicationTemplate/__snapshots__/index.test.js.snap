// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`services.communicationTemplate DB related tests finds the accountTemplate for a given accountId and templateName 1`] = `
Object {
  "accountId": "62954241-3652-4792-bae5-5bfed53d37b7",
  "id": "4486edfc-7e4f-464f-8fe7-a38bb6a37a52",
  "templateId": "1486edfc-7e4f-464f-8fe7-a38bb6a37a52",
  "value": "Personalized template for \${name}, \${age}.",
}
`;

exports[`services.communicationTemplate DB related tests finds the accountTemplate for a given accountId and templateName 2`] = `
Object {
  "id": "1486edfc-7e4f-464f-8fe7-a38bb6a37a52",
  "templateName": "test-template-1",
  "values": Object {
    "age": true,
    "name": true,
  },
}
`;

exports[`services.communicationTemplate DB related tests returns the default accountTemplate if account specific template is not found 1`] = `
Object {
  "accountId": null,
  "id": "2486edfc-7e4f-464f-8fe7-a38bb6a37a52",
  "templateId": "0486edfc-7e4f-464f-8fe7-a38bb6a37a52",
  "value": "Great One \${name}",
}
`;

exports[`services.communicationTemplate DB related tests returns the default accountTemplate if account specific template is not found 2`] = `
Object {
  "id": "0486edfc-7e4f-464f-8fe7-a38bb6a37a52",
  "templateName": "test-template",
  "values": Object {
    "name": true,
  },
}
`;

exports[`services.communicationTemplate template compiler validates template params 1`] = `
Object {
  "age": 22,
  "name": "Donna",
}
`;

exports[`services.communicationTemplate template compiler validates template params for object childs 1`] = `
Object {
  "age": 22,
  "name": "Donna",
  "patient.firstName": "Mike",
}
`;
